clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -5886,16 +5886,71 @@ void CodeGenModule::generateIntelFPGAAnnotation(
   }
 }
 
+/**
+ * Adds global Intel FPGA annotations for a given variable declaration.
+ * This function handles both simple global variables and fields within
+ * structs that are annotated with Intel FPGA attributes. For structs,
+ * it recursively visits all fields and base classes to collect annotations.
+ *
+ * @param VD The variable declaration to annotate.
+ * @param GV The LLVM GlobalValue corresponding to the variable declaration.
+ */
 void CodeGenModule::addGlobalIntelFPGAAnnotation(const VarDecl *VD,
                                                  llvm::GlobalValue *GV) {
+  // A buffer to accumulate annotation strings.
   SmallString<256> AnnotStr;
+
+  // Handle annotations for fields within a device_global struct.
+  // Check if the variable's type is a struct (record type).
+  if (VD->getType()->isRecordType()) {
+    // Retrieve the RecordType from the variable's type.
+    auto RT = VD->getType()->getAs<RecordType>();
+
+    // Define a lambda function for recursive annotation generation.
+    // This lambda will be called for each field and base class in the struct.
+    auto Gen = [&AnnotStr, this](const RecordType *Ty, auto &&Gen) -> void {
+      // Cast the RecordType to a CXXRecordDecl to access fields and bases.
+      const CXXRecordDecl *RD = cast<CXXRecordDecl>(Ty->getDecl());
+
+      // Iterate over the fields of the struct.
+      for (const auto *Field : RD->fields()) {
+	// Generate and append FPGA annotations for the field.
+        generateIntelFPGAAnnotation(Field, AnnotStr);
+
+	// If the field's type is also a struct, recursively call Gen.
+        if (const auto *FT =
+                Field->getType()
+                    ->getPointeeOrArrayElementType() // Strip pointers/arrays
+                    ->getAs<RecordType>())
+          Gen(FT, Gen);
+      }
+
+      // Iterate over the base classes of the struct.
+      for (const auto Base : RD->bases()) {
+	// Retrieve the type of the base class.
+        QualType BaseTy = Base.getType();
+
+	// If the base class type is also a struct, recursively call Gen.
+        if (const auto *BRT = BaseTy->getAs<RecordType>())
+          Gen(BRT, Gen);
+      }
+    };
+    // Start the recursive annotation generation process for the struct.
+    Gen(RT, Gen);
+  }
+
+  // Generate and append FPGA annotations for the global variable itself.
```

Comment:
Just a broad comment overall.

This is probably the right way to implement this (at least I can't think of optimizing this too much yet), but I am concerned that we take a big hit doing all this traversal for every global variable.  Can this be restricted to only when ``-fintelfpga`` is in effect?

Reply:
Thanks @premanandrao for reviews!


>>Can this be restricted to only when -fintelfpga is in effect?

Done

Reply:
@premanandrao, I attempted a different implementation approach, but it was unsuccessful. Thank you to @Fznamznon for the assistance and suggestions. The complexity arises because the header code is intricate, and the code that inspects fields does not reach a field with an attribute. The device_global class in the headers has a base class containing a pointer to a user-defined target struct. Thus, a recursive search through the record's fields and bases is required to strip pointer types.

https://github.com/intel/llvm/blob/f958dce28053dff145cd725ff57bc4ce94cb94d7/sycl/include/sycl/ext/oneapi/device_global/device_global.hpp#L150 
I have updated the patch to restrict the case to only apply when the -fintelfpga flag is active.